pkg/apis/messaging/v1alpha1/in_memory_channel_lifecycle.go

```diff
@@ -0,0 +1,117 @@
+
+const (
+	// InMemoryChannelConditionReady has status True when all subconditions below have been set to True.
+	InMemoryChannelConditionReady = duckv1alpha1.ConditionReady
+
+	// InMemoryChannelConditionDispatcherReady has status True when a Dispatcher deployment is ready
+	// Keyed off appsv1.DeploymentAvaialble, which means minimum available replicas required are up
+	// and running for at least minReadySeconds.
+	InMemoryChannelConditionDispatcherReady duckv1alpha1.ConditionType = "DispatcherReady"
+
+	// InMemoryChannelConditionServiceReady has status True when a k8s Service is ready. This
+	// basically just means it exists because there's no meaningful status in Service. See Endpoints
+	// below.
+	InMemoryChannelConditionServiceReady duckv1alpha1.ConditionType = "ServiceReady"
+
+	// InMemoryChannelConditionEndpointsReady has status True when a k8s Service Endpoints are backed
+	// by at least one endpoint.
+	InMemoryChannelConditionEndpointsReady duckv1alpha1.ConditionType = "EndpointsReady"
+
+	// InMemoryChannelConditionAddressable has status true when this InMemoryChannel meets
+	// the Addressable contract and has a non-empty hostname.
+	InMemoryChannelConditionAddressable duckv1alpha1.ConditionType = "Addressable"
+
+	// InMemoryChannelConditionServiceReady has status True when a k8s Service representing the channel is ready.
+	// Because this uses ExternalName, there are no endpoints to check.
+	InMemoryChannelConditionChannelServiceReady duckv1alpha1.ConditionType = "ChannelServiceReady"
+)
+
+// GetCondition returns the condition currently associated with the given type, or nil.
+func (imcs *InMemoryChannelStatus) GetCondition(t duckv1alpha1.ConditionType) *duckv1alpha1.Condition {
+	return imcCondSet.Manage(imcs).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (imcs *InMemoryChannelStatus) IsReady() bool {
+	return imcCondSet.Manage(imcs).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (imcs *InMemoryChannelStatus) InitializeConditions() {
+	imcCondSet.Manage(imcs).InitializeConditions()
+}
+
+// TODO: Use the new beta duck types.
+func (imcs *InMemoryChannelStatus) SetAddress(hostname string) {
+	imcs.Address.Hostname = hostname
+	if hostname != "" {
+		imcCondSet.Manage(imcs).MarkTrue(InMemoryChannelConditionAddressable)
+	} else {
+		imcCondSet.Manage(imcs).MarkFalse(InMemoryChannelConditionAddressable, "emptyHostname", "hostname is the empty string")
+	}
+}
+
+func (imcs *InMemoryChannelStatus) MarkDispatcherFailed(reason, messageFormat string, messageA ...interface{}) {
+	imcCondSet.Manage(imcs).MarkFalse(InMemoryChannelConditionDispatcherReady, reason, messageFormat, messageA...)
+}
+
+func (imcs *InMemoryChannelStatus) PropagateDispatcherStatus(ds *appsv1.DeploymentStatus) {
+	for _, cond := range ds.Conditions {
```

Comment:
If possible, we should use a common util function for this. https://github.com/knative/eventing/blob/d123d12f7c6b9c51298719c67cd9745571abade2/pkg/apis/eventing/v1alpha1/broker_lifecycle.go#L140